### PR TITLE
Add isless for enum types

### PR DESCRIPTION
**Pull request by kmsquire**
I have a copy of Jameson's original enum.jl in Logging.jl.  While updating for v0.4, I wanted to switch to `Base`'s enum, but found that `isless` wasn't defined (whereas it was in the original).

I know isless might not make sense for many/most enums, but for something like log levels, it does.  No other use cases come to mind, though, so if it's preferrable, I can just define this Logging.jl.  I don't think it hurts anything, though.

Cc: @quinnj @JeffBezanson

### CODE REVIEW

**Comment by pao** [comment]
If this is contentious, alternative idea that can be shot down on its own: `OrderedEnum <: AbstractEnum`, `Enum <: AbstractEnum`?

**Comment by kmsquire** [comment]
> good reasons to shoot this down include not knowing how enums work in the type system

@pao, unclear whether "this" relates to your suggestion or this PR...

At any rate, because `isless` existed in the original implementation and was taken out, there might very well be resistance to this PR.  Making an `@orderedenum` macro should be pretty trivial, though.

```julia
macro orderedenum(T, syms...)
    # copied from @enum
    if isa(T,Symbol)
        typename = T
    elseif isa(T,Expr) && T.head === :curly
        typename = T.args[1]
    else
        throw(ArgumentError("invalid type expression for enum $T"))
    end
    quote
        @enum $T $(syms...)
        Base.isless{E<:$(esc(typename))}(x::E, y::E) = isless(x.val, y.val)
    end
end
```

**Comment by pao** [comment]
Sorry, I did mean my suggestion.

**Comment by StefanKarpinski** [comment]
I'd rather just support this with the normal `@enum` construct and not have too many of these. If you want an unordered enum, don't compare values. There are also situations where it's handy to have *some* ordering to distinct values even if it doesn't matter what that ordering is.

**Comment by StefanKarpinski** [comment]
**tl;dr:** +1 to this PR

**Comment by catawbasam** [comment]
+1

**Comment by tkelman** [comment]
What I don't get is why this would make win64 appveyor segfault, twice now.

**Comment by tkelman** [comment]
From building this branch locally, it looks like maybe it's the same problem that I fixed in #10876 and it might go away with a rebase? Strange since I would expect appveyor to be building a merge commit.

**Comment by pao** [comment]
I suspect the merge ref won't be updated unless the pull request is also updated. If that wasn't the case, your CI would potentially be racing commits to the master branch.

**Comment by tkelman** [comment]
CI does race commits to the master branch. See how https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3970/job/w3e4y9b081jv0nve reported from `versioninfo()` that it was at commit f20c47f, which is a merge into 14-hours-ago master 438e90bb2563d653c0fa6ebee433e3e83836fb60, not 2-days-ago master.

**Comment by pao** [comment]
Well, then I've got nothing. It is definitely fetching the merge ref.

**Comment by kmsquire** [comment]
The 64-bit Windows failure actually shows an [access violation in utf8proc](https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3970/job/rvpm9p7ybxrre7vp#L966).

**Comment by tkelman** [comment]
looks that way, but utf8proc always appears in the backtraces on windows

**Comment by kmsquire** [comment]
I wasn't aware of that.  At any rate, the errors were quite different between 32 and 64 bit, although I guess memory errors can cause that. 

Is this reproducible locally with this branch on Windows?  I guess you've been looking at a lot of Windows failures recently, so not sure if you've looked or have the time to.

**Comment by tkelman** [comment]
I reproduced the problem by checking out this branch, and it went away with a rebase of the branch on top of latest master. I didn't try to checkout out a merge commit locally, and I'm really not sure what the difference would be...

**Comment by pao** [comment]
If you really wanted to know, you could check with `git diff` (no Windows machine required) that the result of fetching `refs/pull/10883/merge` is the same as the rebased branch.

But might be easier to just rebase it, see if AV succeeds, and move on.

**Comment by kmsquire** [comment]
I can certainly rebase and move on.  Since this seemed to be reproducible and might be indicative of a deeper issue, though, I wanted to see if anyone wanted to dig into it deeper before I rebased and overwrote the branch.

Of course, there's always the chance that it was fixed recently on master.

**Comment by tkelman** [comment]
I fetched a merge commit (f54eb2ee42a5e3b7c9d2d060c4eb3a8515cdcb17) locally and it didn't segfault, so I'm satisfied enough with the rebase plan. Wouldn't be the first time that something confusing and unexpected happened on appveyor.

**Comment by kmsquire** [comment]
Rebased and force pushed.  Let's see what appveyor says...

**Comment by pao** [comment]
Well, AppVeyor succeeded!

32-bit dict test failure is not unique to this pull: https://github.com/JuliaLang/julia/issues/9763#issuecomment-94298399
